### PR TITLE
Router02 additions to support FoT tokens

### PR DESCRIPTION
**Pull request by NoahZinsmeister**
stacked on top of #33

modifies `UniswapV2Router02.sol` to add support for standard fee-on-transfer tokens

### CODE REVIEW

**Comment by MicahZoltu** [comment]
Design Opinion: I'm not a fan of inheritance for versioning when you have no ability to actually update previous versions.  That is, you aren't doing something like applying security patches to previous versions and adding features to new versions like in traditional software.  Any changes, including security patches, require a full new version so I don't see the benefit in inheritance here.

**Comment by NoahZinsmeister** [comment]
@moodysalem in particular i'd like your opinion on how the `amounts` array returned from swap functions works now for FoT tokens.

i think it's fairly consistent in that it reports the amount _sent_ to the first pair at the 0 index and the amount _requested_ at all higher indices. this can get a little weird though, as seen here: https://github.com/Uniswap/uniswap-v2-periphery/pull/34/files#diff-a81215b286126f08fb0699e8138282f3R208. basically, the the last element of the array there is the amount of the FoT token that was requested out from the pair, but of course isn't the amount actually received by `to`. another option would be to do the balance math on the `to` address but that feels pretty bad to me.

File: contracts/UniswapV2Router03.sol [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+    {
+        amounts = new uint[](path.length);
+        amounts[0] = amountIn;
+        for (uint i; i < path.length - 1; i++) {
+            (address input, address output) = (path[i], path[i + 1]);
+            IUniswapV2Pair pair = IUniswapV2Pair(UniswapV2Library.pairFor(factory, input, output));
+            uint amountInput;
+            uint amountOutput;
+            { // scope to avoid stack too deep errors
+            (uint reserveInput, uint reserveOutput) = UniswapV2Library.getReserves(factory, input, output);
```
[the remark this answers is not in the document]
Comment (NoahZinsmeister) [reply contracts/UniswapV2Router03.sol:48]: i guess because we're sorting tokens later in the method that's true

**Comment by NoahZinsmeister** [comment]
in my mind the biggest annoyance with separate methods is that rather than simply swapping out router addresses, we'd have to select an entire second set of swapping methods (admittedly with the same arguments, but still a hindrance). this is not a big deal on interfaces, but could be frustrating for smart contract integrations. however, i do agree that a single canonical router is desirable. if the gas costs were just a little less, i would have no problem simply sending non-FoT tokens through the FoT-capable methods (which preserve behavior for non-FoT tokens), but >2k is a bit hefty.

i think that trying to estimate the bips taken off of a FoT token in the registry is way too fraught (i feel like >50% of the FoT tokens i've seen don't adhere to this simplistic fee model). i'd favor a interface which simply calculates trades as if the tokens are not FoT, tries to estimate the gas for the normal method, and if that fails, fall back to trying the FoT-capable method (with perhaps a UI indication). if both fail, there's another issue at play.

w/r/t return values, if we aren't adhering to the V1 interface, i'd prefer to return nothing at all from the FoT methods rather than using balance math any more than strictly necessary. anyone using these methods should know enough to do their own balance checks anyway.

**Comment by moodysalem** [comment]
> in my mind the biggest annoyance with separate methods is that rather than simply swapping out router addresses, we'd have to select an entire second set of swapping methods (admittedly with the same arguments, but still a hindrance). this is not a big deal on interfaces, but could be frustrating for smart contract integrations. however, i do agree that a single canonical router is desirable. if the gas costs were just a little less, i would have no problem simply sending non-FoT tokens through the FoT-capable methods (which preserve behavior for non-FoT tokens), but >2k is a bit hefty.

If the on-chain integrator wishes to support FoT tokens, at least it's an explicit choice in that the method they call is for FoT, rather than one router supporting FoT and the other not, which seems more subtle.

> i think that trying to estimate the bips taken off of a FoT token in the registry is way too fraught (i feel like >50% of the FoT tokens i've seen don't adhere to this simplistic fee model). i'd favor a interface which simply calculates trades as if the tokens are not FoT, tries to estimate the gas for the normal method, and if that fails, fall back to trying the FoT-capable method (with perhaps a UI indication). if both fail, there's another issue at play.

That's great, we avoid the double approve and have no need for the registry.

> w/r/t return values, if we aren't adhering to the V1 interface, i'd prefer to return nothing at all from the FoT methods rather than using balance math any more than strictly necessary. anyone using these methods should know enough to do their own balance checks anyway.

Removing all return arguments SGTM, assuming we still enforce minimumAmountOut, maxmimumAmountIn, though I think it's likely the on-chain caller will almost always want to know the actual amount out and actual amount in respectively

File: contracts/UniswapV2Router03.sol (hunk 1 above)
[the remark this answers is not in the document]
Comment (NoahZinsmeister) [reply contracts/UniswapV2Router03.sol]: changed

**Comment by NoahZinsmeister** [comment]
@moodysalem 3 things in particular to look out for in the latest commit:

1) `removeLiquidityETH{,WithPermit}SupportingFeeOnTransferTokens` still checks slippage parameters against 'ideal' amounts, not post-hoc balances, and now only returns `amountETH`

2) i decided not to add exact output methods that support fee-on-transfer tokens, because it's really impossible to guarantee this property in a generic way. in reality this just means that we need 1 more branch of UI code explaining this

3) all `swap*SupportingFeeOnTransferTokens` functions don't return anything, and check slippage parameters against post-hoc balances

**Review by Rashidsafwan — COMMENTED** [review]
0xC83276021cf27F2C0F7499680f0A868174a20776

**Pull request merged**